cranelift-codegen/meta/src/shared/mod.rs

```diff
@@ -30,21 +35,64 @@ pub(crate) fn define() -> Definitions {
 
     let immediates = Immediates::new();
     let entities = EntityRefs::new();
-    let format_registry = formats::define(&immediates, &entities);
-    let instructions = instructions::define(
-        &mut all_instructions,
-        &format_registry,
-        &immediates,
-        &entities,
-    );
+    let formats = Formats::new(&immediates, &entities);
+    let instructions =
+        instructions::define(&mut all_instructions, &formats, &immediates, &entities);
     let transform_groups = legalize::define(&instructions, &immediates);
 
     Definitions {
         settings: settings::define(),
         all_instructions,
         instructions,
         imm: immediates,
-        format_registry,
+        formats,
         transform_groups,
     }
 }
+
+impl Definitions {
+    /// Verifies certain properties of formats.
+    ///
+    /// - Formats must be uniquely named: if two formats have the same name, they must refer to the
+    /// same data. Otherwise, two format variants in the codegen crate would have the same name.
+    /// - Formats must be structurally different from each other. Otherwise, this would lead to
+    /// code duplicate in the codegen crate.
+    ///
+    /// Returns a list of all the instruction formats effectively used.
+    pub fn verify_instruction_formats(&self) -> Vec<&InstructionFormat> {
```

Comment:
Is this new or does it replace something that I missed?

Reply:
It is replacing code in `lookup`, that made sure to find the right format. Since we're now specifying it manually, a check is required to ensure it matches what the instruction actually needs.